server/src/pull/pull.ts

```diff
@@ -0,0 +1,258 @@
+import {z} from 'zod';
+import type {PatchOperation, PullResponse, PullResponseOKV1} from 'replicache';
+import {transact} from '../pg';
+import {getClientGroupForUpdate, putClientGroup, searchClients} from '../data';
+import type Express from 'express';
+import {hasNextPage, isPageEmpty, readNextPage} from './next-page';
+import {
+  CVR,
+  dropCVREntries,
+  getCVR,
+  putCVR,
+  recordDeletes,
+  recordUpdates,
+} from './cvr';
+import {PartialSyncState, PARTIAL_SYNC_STATE_KEY} from 'shared';
+
+const cookieSchema = z.object({
+  order: z.number(),
+  clientGroupID: z.string(),
+});
+
+export type Cookie = z.infer<typeof cookieSchema>;
+
+export const pullRequest = z.object({
+  clientGroupID: z.string(),
+  cookie: z.union([cookieSchema, z.null()]),
+});
+
+export async function pull(
+  requestBody: Express.Request,
+): Promise<PullResponse> {
+  const start = performance.now();
+  console.log(`Processing pull`, JSON.stringify(requestBody, null, ''));
+  const pull = pullRequest.parse(requestBody);
+  const {clientGroupID} = pull;
+
+  const {clientChanges, nextPage, prevCVR, nextCVR} = await transact(
+    async executor => {
+      // Get a write lock on the client group first to serialize with other
+      // requests from the CG and avoid deadlocks.
+      const baseClientGroupRecord = await getClientGroupForUpdate(
+        executor,
+        clientGroupID,
+      );
+      const prevCVR = pull.cookie
+        ? await getCVR(executor, pull.cookie.clientGroupID, pull.cookie.order)
+        : undefined;
+      const baseCVR: CVR = prevCVR ?? {
+        clientGroupID,
+        clientVersion: 0,
+        order: 0,
+      };
+
+      // Drop any cvr entries greater than the order we received.
+      // Getting an old order from the client means that the client is possibly missing the data
+      // from future orders.
+      //
+      // The other possbility is that multiple tabs are syncing (why are tabs not coordinating who does sync?)
+      // Given that, dropping entries can cause to tabs could starve one another!! I've updated the frontend
```

Comment:
I do not think they can starve each other actually. The snapshots are shared between clients, just slowly (currently ~2s I think). So if you have tab A and B, and A receives order 1, and B 2. Then A pulls again and the server sees that 2 is ahead, it will go ahead and delete and respond to A. But then A will share this snapshot with B and get the new cookie. When B next pulls, he will use the cookie that A received and all is well.

Reply:
See also: https://www.notion.so/replicache/DD-3-1-e42489fc2e6b4340a01c7fa0de353a30

Reply:
> The snapshots are shared between clients, just slowly (currently ~2s I think)

So maybe not starving one another but sounds like there'll be a lot of thrown away syncs while we wait the 2 seconds for the snapshots to get shared.

While snapshot isn't shared:

```
Starting state: server order 2, A order 1, B order 2

A pulls order 1
Server: delete 2, send order 3 to A
B pulls order 2
Server: delete 3, send order 4 to B
A pulls order 3
Server: delete 4, send order 5 to A
```

3 is based on order 1.
4 is based on order 1.
5 is based on order 1.

Nobody made progress in that cycle.

Will take my lock out and test though.

Reply:
Yeah, I agree this is a problem now. A design for the `cvr_entry` table that would be "compatible" with the current design of Replicache would be one where older rows are retained instead of mutated in place. Something like:

```
cvr_entry
=======
client_group_id: int, fk(client_group.id)
tbl: enum
entity_version: int
valid_from_client_view_version: int
valid_to_client_view_version: int nullable # null means "still valid"
```

When an entry in the client view is modified, we:
- set the previous entry's `valid_to_client_view_version`
- create a new version with the appropriate `valid_from_client_view_version`

Alternately, perhaps @grgbkr can remind me one more time why this delay is here. I remember we were trying to ensure that the network sync usually won, but now I can't remember why! Perhaps it would be OK to tweak this so that local sync usually wins. In this case your current design would work.

Reply:
> but sounds like there'll be a lot of thrown away syncs while we wait the 2 seconds for the snapshots to get shared.
...
> 3 is based on order 1.
> 4 is based on order 1.
> 5 is based on order 1.
> Nobody made progress in that cycle.

Wait, I don't agree nobody makes progress.

It's true that in your example the response with order `4` will be based on order `1`. But it's not true that the client requesting it makes no progress. He gets synced to head. It's just that the delta is bigger than necessary: he had up through order `2`, but the patch included those changes even though he didn't need them. This continues with any subsequent pulls until the tabs share their state.

So I think your design is robust without the lock, just slightly inefficient.

This should of course be tested to confirm.